CORE-7793 - Failed serialization when producing Kafka records should send the Flow to the DLQ

pmck91 opened the pull request:
To fix this potential infinite loop, we can return null for un-serializable records, and push an explanation sting to the DLQ as we cant push the data that failed to serialize. 

For other uses of the serializer they may expect the exception, so we only return null in the case a throwOnSerializationError is true.

We have decided that the api for corda's avro serializers should be moved from corda-api to runtime-os
I have ported CordaAvroSerializer, CordaAvroDeserializer, CordaAvroSerializationFactory. 

I have added a Boolean field (defaults to true to match current behaviour) 'throwOnSerializationError ' to the CordaProducerBuilder. this will allow us to decide if we wish to maintain current behaviour and throw when a serialization error occurs, or swallow it and allow it be handled by the onError handler only. 

Most of the files in this PR are just import changes. 
The following are files of interest:

- CordaSubscriptionFactory
- StateAndEventBuilder
- StateAndEventSubscriptionImpl
- CordaAvroSerializer
- CordaAvroSerializerImpl
- CordaAvroDeserializer
- CordaAvroSerializationFactory
- KafkaCordaProducerBuilderImpl

pmck91 committed:
draft pr commit, testing needed

pmck91 committed:
compilation error

pmck91 committed:
Moved CordaAvroSerialzerFactory and Serialzer + Deserializer to runtime OS

LWogan commented on libs/serialization/serialization-avro/src/main/java/net/corda/serialization/package-info.java:
```diff
@@ -0,0 +1,4 @@
+@Export
+package net.corda.serialization;
```
We should move this new lib to a different package other than net.corda.serialization;.  Maybe something with the word avro in it would do?

In OSGI we cant have 2 bundles  export the same package. If we did then  OSGI wouldnt know which bundle to wire to, to get the classes for an imported package. OSGI doesn't look inside the packages for specific classes when wiring together bundles. it just wires import package to export package at startup so it would have a 50/50 chance of getting it right.

Serialization-internal has  already taken net.corda.serialization
https://github.com/corda/corda-runtime-os/blob/release/os/5.0/libs/serialization/serialization-internal/src/main/java/net/corda/serialization/package-info.java

pmck91 committed:
Merge branch 'release/os/5.0' into pmck/CORE-7793

pmck91 committed:
missed some imports, changed package to net.corda.avro.serialization to avoid conflict

pmck91 committed:
missed some imports, changed package to net.corda.avro.serialization to avoid conflict

pmck91 committed:
updating tests to match new factory api

pmck91 committed:
updating tests to match new factory api

pmck91 committed:
updating tests to match new factory api

pmck91 committed:
updating tests to match new factory api

pmck91 committed:
updating tests to match new factory api

LWogan commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddNotaryToGroupParametersHandler.kt:
```diff
@@ -29,7 +29,7 @@ internal class AddNotaryToGroupParametersHandler(
 
     private val keyValuePairListSerializer: CordaAvroSerializer<KeyValuePairList> =
-        cordaAvroSerializationFactory.createAvroSerializer {
+        cordaAvroSerializationFactory.createAvroSerializer(false) {
```
I think this is the only one youve set it to false for. is it an easter egg to see if we're paying attention :D

I think lets just leave it as the default "true" and leave it up to this ticket to modify the logic to handle the error case better https://r3-cev.atlassian.net/browse/CORE-9673

pmck91 replied:
I was searching for this and couldnt find it!

LWogan commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/serialization/CordaAvroSerializerImpl.kt:
```diff
@@ -1,13 +1,16 @@
 import org.apache.kafka.common.serialization.StringSerializer
 import org.slf4j.LoggerFactory
 
 class CordaAvroSerializerImpl<T : Any>(
```
we should put some kdocs and @param tags on this class to help explain the logic

LWogan commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/serialization/CordaAvroSerializerImpl.kt:
```diff
@@ -23,9 +26,21 @@ class CordaAvroSerializerImpl<T : Any>(
-                    log.error("Failed to serialize instance of class type ${data::class.java.name} containing " +
-                            "$data", ex)
-                    throw ex
+                    // We don't want to throw as that would mean the entire poll (with possibly
```
i think move this comment up to the class level. might need a little rewrite. it reads as if this class only handles message pattern serialization, might confuse someone not familiar with the message patterns, consumer and produce loop and that use case etc

LWogan commented on libs/messaging/kafka-message-bus-impl/src/test/kotlin/net/corda/messagebus/kafka/serialization/CordaAvroSerializerImplTest.kt:
```diff
@@ -14,7 +14,7 @@ class CordaAvroSerializerImplTest {
-    private val cordaAvroSerializer = CordaAvroSerializerImpl<Any>(avroSchemaRegistry)
+    private val cordaAvroSerializer = CordaAvroSerializerImpl<Any>(avroSchemaRegistry, true, null)
 
     data class SerializeTester(val contents: String = "test contents")
```
You mentioned you're going to add a unit test, just in case anyone else adds a comment!

LWogan commented on libs/messaging/message-bus/src/main/kotlin/net/corda/messagebus/api/producer/builder/CordaProducerBuilder.kt:
```diff
@@ -13,8 +13,15 @@ interface CordaProducerBuilder {
      * Generate kafka producer with given properties.
      * @param producerConfig The mandatory config for setting up producers
      * @param messageBusConfig Configuration for connecting to the message bus and controlling its behaviour.
+     * @param onSerializationError a callback to receive messages that fail to serialize.  In the producer feed
```
missing kdoc for throwOnError. 

kdoc for onSerializationError might need a little altering as it has the old description about it being used to decide whether to return null

LWogan replied:
This is still a bit off
```
@param onSerializationError a callback to receive messages that fail to serialize. In the producer feed
these will show up as records with a null value, which means they should be removed from any maps.
```

should be something like:
`@param onSerializationError a callback to execute when serialization fails 
`
theres other stuff that uses the serializer, not just producers, and this flag doesnt flag whether we return null anymore

LWogan commented on libs/serialization/serialization-avro/build.gradle:
```diff
@@ -0,0 +1,15 @@
+dependencies {
+    compileOnly "org.osgi:osgi.annotation"
+    compileOnly "org.osgi:org.osgi.service.component.annotations"
+    implementation platform("net.corda:corda-api:$cordaApiVersion")
```
i dont think you need

    implementation platform("net.corda:corda-api:$cordaApiVersion")

LWogan commented on libs/serialization/serialization-avro/src/main/kotlin/net/corda/avro/serialization/CordaAvroSerializationFactory.kt:
```diff
@@ -0,0 +1,22 @@
+
+    /**
+     * Create the {@link CordaAvroSerializer} for use in Avro/message bus serialization.
+     */
```
They were missing already before but we should add some @param and @return  kdocs to explain clearly the new functionality and how it relates to the flags. not sure why we didnt have them before already. 

I know it hasnt changed since the move from corda-api but please add some for the createAvroDeserializer method too.

LWogan commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/serialization/CordaAvroSerializerImpl.kt:
```diff
@@ -23,9 +26,21 @@ class CordaAvroSerializerImpl<T : Any>(
+                    // We don't want to throw as that would mean the entire poll (with possibly
+                    // many records) would fail, and keep failing. So we'll just callback to note the serialize
+                    // and return a null.  This will mean the record gets treated as 'deleted' in the processors
+                    val message = "Failed to serialize instance of class type ${data::class.java.name} containing " +
```
this should fit on one line. i think detekt is configured to 140 line lengths. intellij defaults to 120

LWogan commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/StateAndEventSubscriptionImpl.kt:
```diff
@@ -122,7 +122,10 @@ internal class StateAndEventSubscriptionImpl<K : Any, S : Any, E : Any>(
             try {
                 deadLetterRecords = mutableListOf()
-                nullableProducer = builder.createProducer(config)
+                nullableProducer = builder.createProducer(config, false) { data ->
```
we should add this dead letter records lambda to the same place in  EventLogSubscriptionImpl

LWogan reviewed (commented):
Looks good. Glad to move this out of corda-api given we dont have any intentions on letting customers implement their own avro serializers. These API changes will allow us to better handle these serialization error use cases!

Given this is mostly complete I think we should apply the same changes youve made in CordaAvroSerializerImpl  to the CordaDBAvroSerializerImpl. Should be quick enough and this will keep the logic in our two bus implementations consistent. 

This would allow us to write a db message bus integration test easily to test the dead letter functionality too _I think_. We have some similar tests for the deserialization dead letter logic in testing:message-pattern:integrationTests DurableSubscriptionIntegrationTest & StateAndEventSubscriptionIntegrationTest. We could try out writing a processor that outputs non avro/string/bytearray type objects and see if it will fail serialization and trigger the DLQ.

simon-johnson-r3 commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/builder/StateAndEventBuilder.kt:
```diff
@@ -8,7 +8,12 @@ import net.corda.messaging.subscription.consumer.listener.StateAndEventConsumerR
-    fun createProducer(config: ResolvedSubscriptionConfig): CordaProducer
+    fun createProducer(
+        config: ResolvedSubscriptionConfig,
+        throwOnError: Boolean = true,
```
From this layer of abstraction, and those down except the serializer itself I think `throwOnError` is misleading. I had to go and look what sort of errors it would throw on, and then it wasn't quite clear whether I would get `onSerializationError` called either way, which required another follow through of the code.

I suggest renaming to `throwOnSerializationError` explaining why that is important to control (as per @LWogan's comment above about this information being in the wrong place).

Then I would document that `onSerializationError` is called no matter what `throwOnSerializtionError` is set to.

simon-johnson-r3 commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/serialization/CordaAvroSerializerImpl.kt:
```diff
@@ -23,9 +26,21 @@ class CordaAvroSerializerImpl<T : Any>(
+
+                    if(throwOnError) {
+                        log.error(message, ex)
+                        runOnErrorLambda(message, onError)
```
I can't see a compelling reason not to move `runOnErrorLambda` above the `if`?
Then you can remove the helper method entirely as it's only a one liner anyway.

simon-johnson-r3 commented on libs/serialization/serialization-avro/src/main/kotlin/net/corda/avro/serialization/CordaAvroSerializationFactory.kt:
```diff
@@ -0,0 +1,22 @@
+     * Create the {@link CordaAvroDeserializer} for use in Avro/message bus serialization.
+     */
+    fun <T : Any> createAvroDeserializer(
+        onError: Consumer<ByteArray>,
```
Small point, and not related to this change, but I find it unusual that when I'm creating a deserializer, the first parameter is not the class but the error consumer.

I also find the use of Consumer<> here a little idiosyncratic when this is really an error callback. Consumer has a semantic meaning that is not really appropriate in this use case. Perhaps this is because it was moved from a Java api and some Java functional interface to represent the callback was required then? Either way I think this will confuse future maintainers and accepting a Kotlin lambda would be better and consistent with the other code here.

pmck91 replied:
Ive swapped it to a regular lambda now

simon-johnson-r3 replied:
Am I correct in thinking you switched it on createAvroSerializer but not createAvroDeserializer?

pmck91 replied:
missed it! added now

pmck91 committed:
Code review changes including:

- adding the same logic to DB serializer
- kdocs
- slight refactoring rewrites for clarity

pmck91 committed:
Merge branch 'release/os/5.0' into pmck/CORE-7793

pmck91 committed:
broke compilation by removing api import

pmck91 committed:
serializer test

pmck91 committed:
serializer test

vkolomeyko commented on applications/workers/release/rest-worker/build.gradle:
```diff
@@ -70,6 +70,7 @@ dependencies {
     e2eTestImplementation project(':libs:virtual-node:cpi-upload-endpoints')
     e2eTestImplementation project(':libs:virtual-node:virtual-node-endpoints')
     e2eTestImplementation project(':libs:lifecycle:lifecycle')
+    e2eTestImplementation project(':libs:serialization:serialization-avro')
```
Why do we need this extra "implementation" dependency when no source files been modified which use classes from that module?

pmck91 replied:
without it the task :applications:workers:release:rest-worker:compileE2eTestKotlin fails as it cant find the supertypes which were previously supplied by corda api

```
Supertypes of the following classes cannot be resolved. Please make sure you have the required dependencies in the classpath:
    class net.corda.messagebus.kafka.serialization.CordaAvroSerializationFactoryImpl, unresolved supertypes: net.corda.avro.serialization.CordaAvroSerializationFactory
```

vkolomeyko replied:
But if this is the case should module containing `CordaAvroSerializationFactoryImpl` have dependency on module containing `CordaAvroSerializationFactory` as `api` dependency?
If transitive dependency expressed that way this will eliminate the need to put `implementation project(':libs:serialization:serialization-avro')` in many modules.

Please see my suggestion: https://github.com/corda/corda-runtime-os/pull/3742/files#r1185240153

seanbrereton reviewed (dismissed):
LGTM(build files only)

pmck91 committed:
Merge branch 'release/os/5.0' into pmck/CORE-7793

LWogan commented on libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/serialization/CordaAvroDeserializerImpl.kt:
```diff
@@ -11,7 +11,14 @@ import org.apache.kafka.common.serialization.Deserializer
+ *
+ * @param T Type to deserialize
+ * @property schemaRegistry the Avro-based Schemas
+ * @property onError lambda to be run on deserialization error
```
expectedClass missing

pmck91 committed:
removed consumer, replaced with lambda

pmck91 committed:
Merge branch 'release/os/5.0' into pmck/CORE-7793

LWogan reviewed (commented):
Looking good! Almost there!

pmck91 committed:
code review plus testing import removal from rest worker

pmck91 committed:
Merge branch 'release/os/5.0' into pmck/CORE-7793

pmck91 committed:
re-adding implementation dep to rest worker e2e tests

pmck91 committed:
Merge branch 'release/os/5.0' into pmck/CORE-7793

vkolomeyko commented on libs/messaging/kafka-message-bus-impl/build.gradle:
```diff
@@ -16,6 +16,7 @@ dependencies {
     implementation project(":libs:configuration:configuration-merger")
     implementation project(":libs:metrics")
     implementation project(":libs:utilities")
+    implementation project(":libs:serialization:serialization-avro")
```
```suggestion
    api project(":libs:serialization:serialization-avro")
```
Due to `CordaAvroSerializationFactoryImpl` using `CordaAvroSerializationFactory` as a super-type.

pmck91 committed:
Update libs/messaging/kafka-message-bus-impl/build.gradle

Co-authored-by: Viktor Kolomeyko <viktor.kolomeyko@r3.com>

pmck91 committed:
Merge branch 'release/os/5.0' into pmck/CORE-7793

# Conflicts:
#	components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandler.kt
#	components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandlerTest.kt

pmck91 committed:
refactoring to remove the added boolean flag

pmck91 committed:
detekt

vkolomeyko commented on components/chunking/chunk-db-write-impl/build.gradle:
```diff
@@ -46,6 +46,7 @@ dependencies {
     implementation project(':libs:membership:network-info')
     implementation project(':components:chunking:chunk-db-write')
     implementation project(':components:membership:certificates-service')
+    implementation project(":libs:serialization:serialization-avro")
```
Do we need this line?

LWogan replied:
this one could be changed to integrationTestImplementation i think

vkolomeyko commented on components/flow/flow-p2p-filter-service/build.gradle:
```diff
@@ -13,6 +13,7 @@ dependencies {
     implementation project(":libs:messaging:messaging")
     implementation project(":libs:flows:session-manager")
     implementation project(':libs:utilities')
+    implementation project(":libs:serialization:serialization-avro")
```
Do we need this line?

LWogan replied:
there is code in this lib using this in flow p2p filter

vkolomeyko commented on components/membership/membership-service-impl/build.gradle:
```diff
@@ -27,6 +27,7 @@ dependencies {
     implementation project(":components:membership:registration")
     implementation project(":components:virtual-node:virtual-node-info-read-service")
     implementation project(":components:crypto:crypto-client")
+    implementation project(":libs:serialization:serialization-avro")
```
Do we need this line, provided that no source code been changed?

LWogan replied:
this one is used too

LWogan commented on libs/messaging/message-bus/src/main/kotlin/net/corda/messagebus/api/producer/builder/CordaProducerBuilder.kt:
```diff
@@ -13,8 +13,14 @@ interface CordaProducerBuilder {
      * Generate kafka producer with given properties.
      * @param producerConfig The mandatory config for setting up producers
      * @param messageBusConfig Configuration for connecting to the message bus and controlling its behaviour.
+     * @param throwOnSerializationError throw exception on error or return null defaults to true
```
can remove throwOnSerializationError now

LWogan commented on libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/builder/StateAndEventBuilder.kt:
```diff
@@ -8,7 +8,19 @@ import net.corda.messaging.subscription.consumer.listener.StateAndEventConsumerR
+     * Create producer
+     *
+     * @param config subscription configuration for the messaging layer.
+     * @param throwOnSerializationError throw exception on serialization error, or return null. defaults to throw
```
need to remove throwOnSerializationError kdoc

LWogan commented on libs/serialization/serialization-avro/src/main/kotlin/net/corda/avro/serialization/CordaAvroSerializationFactory.kt:
```diff
@@ -0,0 +1,33 @@
+     * Create the {@link CordaAvroSerializer} for use in Avro/message bus serialization.
+     *
+     * @param T the type to serialize
+     * @param throwOnSerializationError throw an exception on serialization failure, or return null (defaults to true)
```
need to remove throwOnSerializationError kdoc

LWogan reviewed (commented):
couple kdocs to remove, and we need to merge the other pr into this one. Other than that this is good to go

LWogan committed:
CORE-7793 add integration test and handling for serialization errors to db bus producers (#3808)

Add new flag to decide whether or not a producer should throw an error when serialization fails. The only use case we care about not flagging errors is in durable subscriptions where we handle errors using dead letter queues.

Add integration test to validate the DLQ logic.

LWogan committed:
Merge remote-tracking branch 'origin/release/os/5.0' into pmck/CORE-7793

# Conflicts:
#	components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/mgm/MgmRegistrationRequestHandlerTest.kt

LWogan committed:
Merge branch 'pmck/CORE-7793' of https://github.com/corda/corda-runtime-os into pmck/CORE-7793

LWogan committed:
CORE-7793 fix build after merge from release

LWogan committed:
CORE-7793 fix build after merge from release

LWogan committed:
CORE-7793 fix build after merge from release

LWogan committed:
CORE-7793 fix build after merge from release

pmck91 committed:
api v number

pmck91 committed:
Merge remote-tracking branch 'origin/pmck/CORE-7793' into pmck/CORE-7793

LWogan reviewed (dismissed):
LGTM

pmck91 committed:
api v number

charlieR3 reviewed (approved):
LGTM (network team files)

mag1c-48 reviewed (approved):
LGTM - build only
